docs/source/notes/extending.rst

```diff
@@ -1,10 +1,238 @@
+  ...  def __init__(self, N, value):
+  ...      self._N = N
+  ...      self._value = value
+  ...
+  ...  def __repr__(self):
+  ...      return "DiagonalTensor(N={}, value={})".format(self._N, self._value)
+  ...
+  ...  def tensor(self):
+  ...      return self._value * torch.eye(self._N)
+
+This first iteration of the design isn't very useful. The main functionality of
+``ScalarTensor`` is to provide a more compact string representation of a scalar
+tensor than in the base tensor class::
+
+  >>> d = ScalarTensor(5, 2)
+  >>> d
+  ScalarTensor(N=5, value=2)
+  >>> d.tensor()
+  tensor([[2., 0., 0., 0., 0.],
+          [0., 2., 0., 0., 0.],
+          [0., 0., 2., 0., 0.],
+          [0., 0., 0., 2., 0.],
+          [0., 0., 0., 0., 2.]])
+
+If we try to use this object with the :mod:`torch` API, we will run
+into issues::
+
+  >>> import torch
+  >>> torch.mean(d)
+  TypeError: mean(): argument 'input' (position 1) must be Tensor, not ScalarTensor
+
+Adding a ``__torch_function__`` implementation to ``ScalarTensor`` makes it
+possible for the above operation to succeed. Let's re-do our implementation,
+this time adding a ``__torch_function__`` implementation::
+
+  >>> HANDLED_FUNCTIONS = {}
+  >>> class ScalarTensor(object):
+  ...  def __init__(self, N, value):
+  ...      self._N = N
+  ...      self._value = value
+  ...
+  ...  def __repr__(self):
+  ...      return "DiagonalTensor(N={}, value={})".format(self._N, self._value)
+  ...
+  ...  def tensor(self):
+  ...      return self._value * torch.eye(self._N)
+  ...
+  ...  def __torch_function__(self, func, args=(), kwargs=None):
+  ...      if kwargs is None:
+  ...          kwargs = {}
+  ...      if func not in HANDLED_FUNCTIONS:
+  ...          return NotImplemented
+  ...      return HANDLED_FUNCTIONS[func](*args, **kwargs)
+
+The `__torch_function__` method takes three arguments: ``func``, a reference to
+the torch API function that is being overrided, ``args``, the tuple of arguments
+passed to the function, and ``kwargs``, the dict of keyword arguments passed to
+the function. It uses a global dispatch stable named ``HANDLED_FUNCTIONS`` to
+store custom implementations. The keys of this dictionary are functions in the
+``torch`` namepsace and the values are implementations for ``ScalarTensor``.
+
+This class definition isn't quite enough to make ``torch.mean`` do the right
+thing when we pass it a ``ScalarTensor`` -- we also need to define an
+implementation for ``torch.mean`` for ``ScalarTensor`` operands and add the
+implementation to the ``HANDLED_FUNCTIONS`` dispatch table dictionary. One way
+of doing this is to define a decorator::
+
+  >>> import functools
+  >>> def implements(torch_function):
+  ...     """Register a torch function override for ScalarTensor"""
+  ...     @functools.wraps(torch_function)
+  ...     def decorator(func):
+  ...         HANDLED_FUNCTIONS[torch_function] = func
+  ...         return func
+  ...     return decorator
```

Comment:
On the other hand, if `HANDLED_FUNCTIONS` pattern takes this much boilerplate, maybe it's a good idea to enshrine this as part of the API, so people don't have to copy paste this code into every example they want to do. (BTW, this is not written in a very copy-pasteable way...)